# Exercise 4.5: Clean Products.csv dataset

# Import Libraries

In [1]:
import pandas as pd
import numpy as np
import os

# Import Dataset

In [2]:
path = r'C:\Users\sands\Documents\01-2022 Instacart Basket Analysis'

In [3]:
df_prods = pd.read_csv(os.path.join(path,'02 Data','Original Data','products.csv'), index_col=False)

In [4]:
df_ords = pd.read_csv(os.path.join(path,'02 Data','Prepared Data','orders_wrangled.csv'),index_col=False)

# Data Consistency Checks

In [5]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       order_day_of_week  order_hour_of_day  days_since_prior_order  
count       3.421083e+06       3.421083e+06            3.214874e+06  
mean        2.776219e+00       1.345202e+01            1.111484e+01  
std         2.046829e+00       4.226088e+00            9.206737e+00  
min         0.000000e+00       0.000000e+00            0.000000e+00  
25%         1.000000e+00       1.000000e+01            4.000000e+00  
50%         3.000000e+00       1.300000e+01            7.000000e+00  
75%         5.000000e+00       1.600000e+01            1.500000e+01  
max         6.000000e+00       2.300000e+01            3.000000e+01

## Create a new dataframe

In [6]:
#Create a new dataset
df_test = pd.DataFrame()

In [7]:
#Assign mixed values to it
df_test['mix'] = ['a','b','1','True']

In [8]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [9]:
#Check for mixed data type

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

In [10]:
#Change the data type of the column

df_test['mix'] = df_test['mix'].astype('str')

In [11]:
#Check for Null values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [12]:
#Create a subset of these null values in the product_name column
df_nan = df_prods[df_prods['product_name'].isnull()==True]

In [13]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [14]:
#Find the mean using df.describe function
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [15]:
#Impute the missing values with the mean
# df['column with missings'].fillna(mean value, inplace=True)

In [16]:
#Find the median
df_prods.median()

C:\Users\sands\AppData\Local\Temp\ipykernel_5896\1301196815.py:2: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  df_prods.median()


product_id       24845.0
aisle_id            69.0
department_id       13.0
prices               7.1
dtype: float64

In [17]:
#Impute the missing values with the median
# df['column with missings'].fillna(median value, inplace=True)

In [18]:
df_prods.shape

(49693, 5)

In [19]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [20]:
#Check the size of the subset and make sure the number of rows has reduced
df_prods_clean.shape

(49677, 5)

In [21]:
#Drop the missing value rows
df_prods.dropna(subset = ['product_name'], inplace = True)

In [22]:
#Check the size of the dataframe and make sure the number of rows has reduced
df_prods.shape

(49677, 5)

In [23]:
#Find full duplicates in the dataset
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [24]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [25]:
#Check the number of rows before deleting the duplicates
df_prods_clean.shape

(49677, 5)

In [26]:
# Create a new dataframe without the duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [27]:
#Check the size of the new dataframe
df_prods_clean_no_dups.shape

(49672, 5)

In [28]:
#Export the cleaned Products dataframe
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# Answers 4.5: Clean Orders.csv Dataset

In [29]:
# Answer 2: Run the describe function on the dataset
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       order_day_of_week  order_hour_of_day  days_since_prior_order  
count       3.421083e+06       3.421083e+06            3.214874e+06  
mean        2.776219e+00       1.345202e+01            1.111484e+01  
std         2.046829e+00       4.226088e+00            9.206737e+00  
min         0.000000e+00       0.000000e+00            0.000000e+00  
25%         1.000000e+00       1.000000e+01            4.000000e+00  
50%         3.000000e+00       1.300000e+01            7.000000e+00  
75%         5.000000e+00       1.600000e+01            1.500000e+01  
max         6.000000e+00       2.300000e+01            3.000000e+01

### The values are resonable and don't need furthur investigation 

In [30]:
# Answer 3: Check for Mixed-type Data in orders.csv dataset
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

### There are no mixed value columns

In [31]:
# Answer 5: Run a check for missing values in your df_ords dataframe
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_day_of_week              0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

### Only the days_since_prior_order column has nulls. The max value for user_id is 206,209 and same are the NaN values.

In [32]:
# Answer 6: Create a new dataframe
df_ords_cleaned = df_ords

In [33]:
#Add a new new column to identify the new customers
df_ords_cleaned['new_customer'] = df_ords['days_since_prior_order'].isnull() == True

In [34]:
df_ords_cleaned

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
1                 1   2398795        1    prior             2   
2                 2    473747        1    prior             3   
3                 3   2254736        1    prior             4   
4                 4    431534        1    prior             5   
...             ...       ...      ...      ...           ...   
3421078     3421078   2266710   206209    prior            10   
3421079     3421079   1854736   206209    prior            11   
3421080     3421080    626363   206209    prior            12   
3421081     3421081   2977660   206209    prior            13   
3421082     3421082    272231   206209    train            14   

         order_day_of_week  order_hour_of_day  days_since_prior_order  \
0                        2                  8                     NaN   
1                        3                  7                    15.0   
2                        3                 12                    21.0   
3                        4                  7                    29.0   
4                        4                 15                    28.0   
...                    ...                ...                     ...   
3421078                  5                 18                    29.0   
3421079                  4                 10                    30.0   
3421080                  1                 12                    18.0   
3421081                  1                 12                     7.0   
3421082                  6                 14                    30.0   

         new_customer  
0                True  
1               False  
2               False  
3               False  
4               False  
...               ...  
3421078         False  
3421079         False  
3421080         False  
3421081         False  
3421082         False  

[3421083 rows x 9 columns]

### Created a new dataframe and added a column new customer to identify if the customer is new or old based on weather the says_since_prior_order is Nan

In [35]:
# Answer 7: Run a check for duplicate values in your df_ords data
df_ords_dups = df_ords_cleaned[df_ords_cleaned.duplicated()]

In [36]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, order_day_of_week, order_hour_of_day, days_since_prior_order, new_customer]
Index: []

### No duplicates were found

In [37]:
# Answer 9: Export the cleaned dataset to the Prepared folder
df_ords_cleaned.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))